artifactory.py

```diff
@@ -1015,14 +1036,50 @@ def del_properties(self, pathobj, props, recursive):
     def scandir(self, pathobj):
         return _ScandirIter((pathobj.joinpath(x) for x in self.listdir(pathobj)))
 
-    def writeto(self, fd, out, chunk_size):
-        url = str(fd)
-        res = fd.session.get(url, stream=True, verify=True, cert=None)
-        if res.status_code != 200:
-            raise RuntimeError(res.status_code)
-        for chunk in res.iter_content(chunk_size=chunk_size):
-            if chunk:
-                out.write(chunk)
+    def writeto(
+        self,
+        arti_cls,
+        output,
+        chunk_size=1024,
+        progress_func=print_download_progress,
+        *args,
+        **kwargs
+    ):
+        """
+        Downloads large file in chunks and prints progress
+        :param arti_cls: ArtifactoryPath class
+        :param output: file path of output file
+        :param chunk_size: chunk size in bytes, recommend. eg 1024*1024 is 1Mb
+        :param progress_func: custom function to print output, otherwise uses default print_download_progress
+        :param args: *args to supply for your custom progress_func
+        :param kwargs: **kargs to supply for your custom progress_func
+        :return: None
+        """
+        response = self.get_response(arti_cls)
+
+        bytes_read = 0
+        if "Content-Length" in response.headers:
+            file_size = int(response.headers["Content-Length"])
+        else:
+            file_size = 0  # sometimes response does not have content length
+
+        if isinstance(output, str) or isinstance(output, pathlib.Path):
+            # if string or Path is provided as output, then open file to write in bytes
+            file = open(output, "wb")
+        else:
+            # file object: TextIOWrapper
+            file = output
+
+        try:
+            for chunk in response.iter_content(chunk_size=chunk_size):
+                bytes_read += len(chunk)
+                if callable(progress_func):
+                    progress_func(bytes_read, file_size, *args, **kwargs)
+                file.write(chunk)
+        finally:
+            # in case if connection will die during download
+            if isinstance(output, str) or isinstance(output, pathlib.Path):
```

Comment:
`try\finally` should be in the `wtireto` below as well. 
Could we replace try\finally solution with `nullcontext` from https://docs.python.org/3/library/contextlib.html?

```python
if isinstance(output, str) or isinstance(output, pathlib.Path):
    context = nullcontext()
else:
   context = open(output, "wb")
```

Reply:
I think it should be other way, but I was looking for this nullcontext and could not find it, thanks for hint!

Reply:
@allburov 
null context is not available in 3.5. Do we consider to upgrade package?
EOL for 3.5 was (13 Sep 2020)

we can consider 3.7.9

Reply:
@allburov 
ou, it looks it is even not in Python 3.6

so here are 3 possible scenarios:
1. just copy class definition into tools/contextlib.py
2. move to python 3.7.9 as minimum version
3. use try/finally

I personally is for either 1 or 2

Reply:
Oh, it's my fault, sorry.
Let's copy definition just here (don't create a file) and document why we did it this way!